sandbox/rocky/tf/algos/batch_polopt.py

```diff
@@ -103,35 +104,41 @@ def train(self, sess=None):
-            
-        sess.run(tf.global_variables_initializer())
-        self.start_worker()
-        start_time = time.time()
-        for itr in range(self.start_itr, self.n_itr):
-            itr_start_time = time.time()
-            with logger.prefix('itr #%d | ' % itr):
-                logger.log("Obtaining samples...")
-                paths = self.obtain_samples(itr)
-                logger.log("Processing samples...")
-                samples_data = self.process_samples(itr, paths)
-                logger.log("Logging diagnostics...")
-                self.log_diagnostics(paths)
-                logger.log("Optimizing policy...")
-                self.optimize_policy(itr, samples_data)
-                logger.log("Saving snapshot...")
-                params = self.get_itr_snapshot(itr, samples_data)  # , **kwargs)
-                if self.store_paths:
-                    params["paths"] = samples_data["paths"]
-                logger.save_itr_params(itr, params)
-                logger.log("Saved")
-                logger.record_tabular('Time', time.time() - start_time)
-                logger.record_tabular('ItrTime', time.time() - itr_start_time)
-                logger.dump_tabular(with_prefix=False)
-                if self.plot:
-                    rollout(self.env, self.policy, animated=True, max_path_length=self.max_path_length)
-                    if self.pause_for_plot:
-                        input("Plotting evaluation run: Press Enter to "
-                              "continue...")
+        try:    
+            sess.run(tf.global_variables_initializer())
+            self.start_worker()
+            start_time = time.time()
+            initialize_plot = 0
+            for itr in range(self.start_itr, self.n_itr):
+                itr_start_time = time.time()
+                with logger.prefix('itr #%d | ' % itr):
+                    logger.log("Obtaining samples...")
+                    paths = self.obtain_samples(itr)
+                    logger.log("Processing samples...")
+                    samples_data = self.process_samples(itr, paths)
+                    logger.log("Logging diagnostics...")
+                    self.log_diagnostics(paths)
+                    logger.log("Optimizing policy...")
+                    self.optimize_policy(itr, samples_data)
+                    logger.log("Saving snapshot...")
+                    params = self.get_itr_snapshot(itr, samples_data)  # , **kwargs)
+                    if self.store_paths:
+                        params["paths"] = samples_data["paths"]
+                    logger.save_itr_params(itr, params)
+                    logger.log("Saved")
+                    logger.record_tabular('Time', time.time() - start_time)
+                    logger.record_tabular('ItrTime', time.time() - itr_start_time)
+                    logger.dump_tabular(with_prefix=False)
+                    if self.plot:
+                        if initialize_plot == 0:
```

Comment:
you can eliminate this state machine if you call plotter.init_plot() outside the for loop. does that not work?

Reply:
When the thread is created outside for loop, the sample generation takes a huge amount of time (~15 mins) on my setup. I believe that the newly created thread is consuming a lot of processor time. To overcome this issue, I added it just before plotting.

Reply:
That may fix your problem in this case, but managing such behavior is an unreasonable expectation on someone trying to use this plotter API in the future. It needs to be clean and manage itself.

If I start an algorithm from scratch, how am I supposed to know I have to initialize the plotter in a specific place in my code?

Did you find the root cause of this problem?

Reply:
Nope. I am not sure why this behavior. Only reason I suspect is because of the while loop in plotter thread.